rust/actyx/swarm/harness/src/bin/versions.rs

```diff
@@ -149,100 +159,106 @@ fn main() -> anyhow::Result<()> {
-                        }],
-                    })
-                    .await
-                })
-                .await?;
-            tracing::info!("{} published: {:?}", i, r);
-        }
+    pub async fn run(tmp: &Path) -> anyhow::Result<()> {
+        let mut sim = Netsim::<Command, Event>::new();
+        spawn_network(&mut sim, tmp).await?;
+        let machines = sim.machines().iter().map(|m| m.id()).collect::<Vec<_>>();
+        ensure_all_machines_are_up(&mut sim, &machines).await?;
 
-        let started = Instant::now();
+        let api = Api::with_port(&mut sim, app_manifest(), 4454).context("creating Api")?;
+        let mut streams = HashMap::new();
         for i in &machines {
-            loop {
-                let v = api
-                    .run(*i, |api| async move {
-                        api.query(QueryRequest {
-                            lower_bound: None,
-                            upper_bound: None,
-                            query: "FROM 'version2'".parse()?,
-                            order: Order::Asc,
-                        })
-                        .await
-                    })
-                    .await?
-                    .filter_map(|r| {
-                        tracing::info!("{} query response {:?}", i, r);
-                        match r {
-                            QueryResponse::Event(EventResponse {
-                                meta: EventMeta::Event { key, .. },
-                                ..
-                            }) => ready(streams.get(&key.stream)),
-                            _ => ready(None),
-                        }
-                    })
-                    .collect::<HashSet<MachineId>>()
-                    .await;
-                if v == all {
-                    break;
-                }
-                if started.elapsed() > Duration::from_secs(1) {
-                    anyhow::bail!("timeout waiting for event propagation");
-                }
-                sleep(Duration::from_millis(100));
-            }
+            let node_id = api.run(*i, |api| async move { Ok(api.node_id().await) }).await?;
+            streams.insert(node_id.stream(0.into()), *i);
         }
 
+        // Step 1: wait until events propagate, i.e. all are connected
+        // We're using a total timeout of 20 seconds and an interval of 1 second between queries
+        // These values date back to before the refactor
+        publish_and_query(
+            &api,
+            &machines,
+            &streams,
+            "versions",
+            Duration::from_secs(20),
+            Duration::from_secs(1),
+        )
+        .await?;
+
+        // it may be that the above was fulfilled by indirect event delivery, but now every
+        // node must see every other; just wait a second to get them connected
+        async_std::task::sleep(Duration::from_secs(1)).await;
```

Comment:
curious, is this a similar attempt to fully_meshed?

Reply:
Good question, I would need to read it to know 😅